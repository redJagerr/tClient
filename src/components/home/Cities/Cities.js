import { useEffect } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { fetchCities } from 'redux/slices/citiesSlice';
import { selectCities } from 'redux/selectors/globalSelectors';
import CityCard from '../CityCard/CityCard';
import CityCardSkeleton from '../CityCard/CityCardSkeleton';
import { v4 } from 'uuid';
import declarationOfNumber from 'utils/declarationOfNumber';

const Cities = () => {
	const { items, status, isSearch } = useSelector(selectCities);
	const dispatch = useDispatch();
	useEffect(() => {
		dispatch(fetchCities());
	}, []);

	const citiesRender = arr =>
		arr.map(({ _id, ...props }) => <CityCard key={_id} id={_id} {...props} />);

	const elements = citiesRender(items);
	return (
		<section className='cities'>
			<h3 className='title-30 cities__title'>
				{isSearch
					? `Найден${declarationOfNumber(items.length, ['', 'о', 'о'])}: ${
							items.length
					  } ${declarationOfNumber(items.length, [
							'город',
							'города',
							'городов',
					  ])}`
					: 'Популярные города'}
			</h3>
			<ul className='list cities__list' data-testid='list'>
				{status === 'loading'
					? [
							<CityCardSkeleton key={v4()} />,
							<CityCardSkeleton key={v4()} />,
							<CityCardSkeleton key={v4()} />,
					  ]
					: elements}
			</ul>
		</section>
	);
};
export default Cities;
