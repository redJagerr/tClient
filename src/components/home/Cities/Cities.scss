.cities {
	margin-top: toRem(30);
	&__title {
		font-size: 30px;
	}
	&__list {
		display: grid;
		grid-template-columns: repeat(3, 1fr);
		grid-auto-rows: minmax(auto, toRem(290));
		gap: 18px;
	}
	&__item {
		background-color: var(--color-gray);
		overflow: hidden;
		position: relative;
		border-radius: 20px;
		transition: all 0.6s ease;
		&:hover {
			transform: translateY(-10px);
		}
	}
	&__link {
		position: absolute;
		top: 0;
		left: 0;
		display: block;
		min-width: 100%;
		height: 100%;
	}
	&__button {
		position: absolute;
		top: toRem(12);
		right: toRem(12);
	}
	&__text {
		position: absolute;
		bottom: 0;
		left: 0;
		color: white;
		padding-left: toRem(28);
		padding-bottom: toRem(28);
		display: flex;
		flex-direction: column;
		z-index: 15;
		&:not(:last-child) {
			margin-bottom: toRem(6);
		}
	}
	&__name {
		font-size: toRem(24);
		font-weight: 500;
	}
	&__count {
		font-size: toRem(12);
		font-weight: 400;
	}
	&__skeleton {
		position: relative;
		display: flex;
		flex-direction: column;
		justify-content: flex-end;
		width: 100%;
		height: 100%;
		background-color: var(--color-gray);
		padding: toRem(32) toRem(28);
		border-radius: toRem(20);
	}
	&__icon {
		width: toRem(32);
		height: toRem(32);
		color: var(--color-gray);
		&:hover {
			color: var(--color-red);
		}
	}
	@include laptop {
		&__list {
			grid-template-columns: repeat(2, 1fr);
			grid-auto-rows: minmax(auto, toRem(230));
		}
	}
	@include tablet {
		&__title {
			font-size: 4vw;
		}
		&__text {
			padding-left: toRem(18);
			padding-bottom: toRem(18);
		}
	}
	@include mobile {
		&__list {
			grid-auto-rows: minmax(auto, toRem(160));
		}
	}
}
.like {
	fill: rgb(131, 131, 131);
	&:hover {
		fill: #ed7c7c;
	}
}
