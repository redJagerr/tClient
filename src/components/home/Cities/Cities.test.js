import * as reduxHooks from 'react-redux';
import Cities from './Cities';
import { renderWithRoutes } from 'tests/helpers/renderWithRoutes';

jest.mock('react-redux');
const useStateCB = {
	items: [
		{
			_id: 'id',
			name: 'Краснодар',
			places: [],
			districts: [],
			photos: '',
		},
	],
	isSearch: false,
	loadStatus: 'idle',
};
const useStateEmptyCB = {
	items: [],
	isSearch: false,
	loadStatus: 'idle',
};

const mockUseDispatch = jest.spyOn(reduxHooks, 'useDispatch');
const mockUseSelector = jest.spyOn(reduxHooks, 'useSelector');
const dispatch = jest.fn();
describe('Cities', () => {
	it('should create Cities with city items array ', () => {
		mockUseSelector.mockReturnValue(useStateEmptyCB);
		mockUseDispatch.mockReturnValue(dispatch);
		const component = renderWithRoutes(<Cities />);
		expect(component).toMatchSnapshot();
	});
	it('should create Cities with empty city items array ', () => {
		mockUseSelector.mockReturnValue(useStateCB);
		mockUseDispatch.mockReturnValue(dispatch);
		const component = renderWithRoutes(<Cities />);
		expect(component).toMatchSnapshot();
	});
});
