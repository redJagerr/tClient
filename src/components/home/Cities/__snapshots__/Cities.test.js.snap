// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Cities should create Cities with city items array  1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <section
        class="cities"
      >
        <h3
          class="title-30 cities__title"
        >
          Популярные города
        </h3>
        <ul
          class="list cities__list"
          data-testid="list"
        />
      </section>
    </div>
  </body>,
  "container": <div>
    <section
      class="cities"
    >
      <h3
        class="title-30 cities__title"
      >
        Популярные города
      </h3>
      <ul
        class="list cities__list"
        data-testid="list"
      />
    </section>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`Cities should create Cities with empty city items array  1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <section
        class="cities"
      >
        <h3
          class="title-30 cities__title"
        >
          Популярные города
        </h3>
        <ul
          class="list cities__list"
          data-testid="list"
        >
          <li
            class="cities__item"
          >
            <a
              class="link cities__link"
              href="/id/places"
              tabindex="0"
            >
              <div
                class="cities__text"
              >
                <span
                  class="cities__name"
                >
                  Краснодар
                </span>
                <span
                  class="cities__count"
                >
                  0
                   
                  интересных мест
                </span>
              </div>
              <img
                alt="Moscow"
                class="image dark-bg"
                src=""
              />
            </a>
          </li>
        </ul>
      </section>
    </div>
  </body>,
  "container": <div>
    <section
      class="cities"
    >
      <h3
        class="title-30 cities__title"
      >
        Популярные города
      </h3>
      <ul
        class="list cities__list"
        data-testid="list"
      >
        <li
          class="cities__item"
        >
          <a
            class="link cities__link"
            href="/id/places"
            tabindex="0"
          >
            <div
              class="cities__text"
            >
              <span
                class="cities__name"
              >
                Краснодар
              </span>
              <span
                class="cities__count"
              >
                0
                 
                интересных мест
              </span>
            </div>
            <img
              alt="Moscow"
              class="image dark-bg"
              src=""
            />
          </a>
        </li>
      </ul>
    </section>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
